Add `maximumDepthExceeded` result kind to getReferencedGlobalParams

maxburs opened the pull request:
getReferencedGlobalParams can fail. Currently, it does so silently. PR updates it's returned values to include an error so we can let users know when it's failing.

I considered making diagnostics and parameters 2 properties instead of using a union, but, it seemed unlikely callers would know what to do with the raw list, and as far as I know, this is the only one they might care about.

Bug: https://msazure.visualstudio.com/One/_workitems/edit/29324049

morgilad commented:
@maxburs please share some more info on what you did and the motivation.
If a screenshot is relevant - it can also help.

morgilad commented on package/tests/integration/api.spec.ts:
```diff
@@ -0,0 +1,26 @@
+import { createMonaKustoModel, MonaKustoModel, loadPageAndWait } from './testkit';
+import { KustoWorker } from '../../src/types';
+
+test.describe('getReferencedGlobalParams', () => {
```
We don’t test individual functions in integration tests. we test functionality.
Please update the suite and test description to explain what functionality is being tested, rather than specifying the function itself.

sagivf commented:
@maxburs this was the reason for the PR right?
https://msazure.visualstudio.com/One/_workitems/edit/29324049

Can you add it to the PR desciription along with some TLDR description?

sagivf commented on package/package.json:
```diff
@@ -1,6 +1,6 @@
 {
     "name": "@kusto/monaco-kusto",
-    "version": "13.1.1",
+    "version": "14.0.0",
```
I think becuase it changes the api of getReferencedGlobalParams which makes it a breaking change.
 We could add a new method to avoid it, but since it's a clear upgrade, including types, I think this is better.

maxburs replied:
\+ 1 to what Sagiv said. This is a breaking change, so, a major version bump is required.

maxburs commented:
> @maxburs please share some more info on what you did and the motivation. If a screenshot is relevant - it can also help.

Updated

Teams post has more information. There isn't any UI for this yet.

> @maxburs this was the reason for the PR right? https://msazure.visualstudio.com/One/_workitems/edit/29324049
> 
> Can you add it to the PR desciription along with some TLDR description?

Added

Do we want to include references to internal stuff on github? I included more information on the Teams post.

maxburs commented on package/tests/integration/api.spec.ts:
```diff
@@ -0,0 +1,26 @@
+                const worker: KustoWorker = await (window as any).getKustoWorker();
+                return await worker.getReferencedGlobalParams((window as any).editor.getModel().uri.toString(), 0);
+            });
+        };
```
I agree a Jest test makes more sense here because it's not ui-centric, but, I'm not sure we're aligned on how should/can be written with jest.

- To effectively test this, we have to initialize everything—monaco editor and the worker. Anything less would make the test harder to setup and not representative of how it's actually used.
- The repo currently doesn't contain factories for setting up Jest tests for things like this

I really want to avoid creating a lot of new test infrastructure that will end up creating a lot of back-and-forth in the PR. I know Mor has strong opinions on how this stuff should be written, so, I'd prefer to avoid trying to guess what people want.

> This should just be a unit test or (and) an integration test - add a button to the testing environment that causes the getReferencedParameters to run and display an error in this case, which would be validated in the test.

I could do this, but, I'm not sure I understand the motivation. The goal is to how a host app might use this worker functionality—aliasing this through new UX will make it less representative of how it's used.

morgilad replied:
As the owner of this codebase and the person responsible for maintaining it, I can’t approve a test I don’t understand. Here’s how you can make it clearer:

Suite Description
It shouldn’t just be the function name. What product area is this test covering? For example: Dashboard parameters validation (open to your suggestions). Also, the test file should match the suite name. 

Test Description
Avoid using only the error name. Explain what the test verifies. For example: Displays an error when query parameters exceed size limits.

Test Logic
Instead of exposing an internal function via evaluate, add a UI-level indicator in the test environment for when a parameter-related error occurs and assert that. This keeps the test aligned with real usage rather than internal implementation details.

Yes, I have strong opinions on test readability because they need to be maintainable long-term. I also receive feedback on my PRs, so this isn’t about avoiding discussion, it’s about maintaining code quality and clarity.

sagivf replied:
@maxburs this isn't a question of style. While exposing an internal api in tests and evaluating is not unheard of, it's not the ideal way to go and is a lot less easy to reason about.

Unless I'm missing something, the solution here is super simple. Add code to the test page that handles kind  "maximumDepthExceeded" and show a message bar with a message, then run the query in the test and validate you get the error message. That avoids the need of a unit test IMO. Am I missing something?

maxburs replied:
@sagivf  @morgilad The test now triggers `getReferencedGlobalParams` through a button:

<img width="585" height="176" alt="image" src="https://github.com/user-attachments/assets/838ef951-02bc-4b88-8fcf-edb6ac8882f7" />

<img width="546" height="171" alt="image" src="https://github.com/user-attachments/assets/36eeb7f2-9e9f-488a-bd2c-87053cd4f750" />
